File: scraper/bills/mongo.py
from datetime import datetime
from pymongo import MongoClient
from bills.bill import Bill


class Mongo:
    def __init__(self, conn_str: str, db: str, col: str) -> None:
        self.client = MongoClient(conn_str)
        self.db = self.client[db]
        self.col = self.db[col]

    def bill_exists(self, bill_id: str, last_updated: datetime) -> bool:
        bill = self.col.find_one({"bill_id": bill_id, "updated": last_updated})
        return bill is not None

    def upsert_bill(self, bill: Bill):
        query = {"bill_id": bill.bill_id}
        update = {
            "$set": {
                "bill_id": bill.bill_id,
                "title": bill.title,
                "short_title": bill.short_title,
                "url": bill.url,
                "text": bill.text,
                "version": bill.version,
                "type": bill.type,
                "issued": bill.issued,
                "updated": bill.updated,
                "actions": [{"text": a.text, "date": a.date} for a in bill.actions],
            }
        }
        self.col.update_one(query, update, upsert=True)

File: monitor/bills_us/mongo.go
package bills_us

import (
	"context"
	"log"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type BillActions struct {
	Date   time.Time `bson:"date"`
	Action string    `bson:"action"`
}

type BillMembers struct {
	Role    string `bson:"role"`
	Chamber string `bson:"chamber"`
	BioID   string `bson:"bio_id"`
	Name    string `bson:"name"`
	State   string `bson:"state"`
	Party   string `bson:"party"`
}

type BillCommittees struct {
	AuthorityID string `bson:"authority_id"`
	Chamber     string `bson:"chamber"`
	Name        string `bson:"name"`
	Type        string `bson:"type"`
}

type Bill struct {
	ID          primitive.ObjectID `bson:"_id"`
	BillID      string             `bson:"bill_id"`
	ShortTitles []string           `bson:"short_titles"`
	Title       string             `bson:"title"`
	Url         string             `bson:"url"`
	Text        string             `bson:"text"`
	Congress    string             `bson:"congress"`
	Session     string             `bson:"session"`
	Pages       int                `bson:"pages"`
	Version     string             `bson:"version"`
	Type        string             `bson:"type"`
	Members     []BillMembers      `bson:"members"`
	Committees  []BillCommittees   `bson:"committees"`
	Actions     []BillActions      `bson:"actions"`
	PublishedAt time.Time          `bson:"published_at"`
	UpdatedAt   time.Time          `bson:"updated_at"`
}

type Mongo struct {
	client *mongo.Client
	col    *mongo.Collection
}

func NewMongo(url, db string, col string) *Mongo {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	client, err := mongo.Connect(ctx, options.Client().ApplyURI(url))
	if err != nil {
		log.Fatalf("could not connect to Mongo: %v", err)
	}
	log.Printf("successfully connected to Mongo")

	coll := client.Database(db).Collection(col)
	return &Mongo{client: client, col: coll}
}

func (m *Mongo) Disconnect(ctx context.Context) error {
	return m.client.Disconnect(ctx)
}

func (m *Mongo) GetBillById(ctx context.Context, packageId string) (*Bill, error) {
	var bill Bill
	err := m.col.FindOne(ctx, bson.D{{Key: "bill_id", Value: packageId}}).Decode(&bill)
	if err != nil {
		return nil, err
	}
	return &bill, nil
}

func (m *Mongo) InsertBill(ctx context.Context, bill *Bill) error {
	bill.ID = primitive.NewObjectID()
	_, err := m.col.InsertOne(ctx, bill)
	return err
}

func (m *Mongo) UpdateBill(ctx context.Context, bill *Bill) error {
	_, err := m.col.UpdateOne(ctx, bson.D{{Key: "_id", Value: bill.ID}}, bson.D{{Key: "$set", Value: bill}})
	return err
}
